patches/minecraft/net/minecraft/world/gen/feature/structure/VillagePieces.java.patch

```diff
@@ -0,0 +1,101 @@
+       Iterator<VillagePieces.PieceWeight> iterator = list.iterator();
+ 
+       while(iterator.hasNext()) {
+@@ -111,6 +112,8 @@
+          villagepieces$village = VillagePieces.House2.func_175855_a(p_176065_0_, p_176065_2_, p_176065_3_, p_176065_4_, p_176065_5_, p_176065_6_, p_176065_7_, p_176065_8_);
+       } else if (oclass == VillagePieces.House3.class) {
+          villagepieces$village = VillagePieces.House3.func_175849_a(p_176065_0_, p_176065_2_, p_176065_3_, p_176065_4_, p_176065_5_, p_176065_6_, p_176065_7_, p_176065_8_);
++      } else {
++         villagepieces$village = net.minecraftforge.fml.common.registry.VillagerRegistry.getVillageComponent(p_176065_1_, p_176065_0_, p_176065_2_, p_176065_3_, p_176065_4_, p_176065_5_, p_176065_6_, p_176065_7_, p_176065_8_);
+       }
+ 
+       return villagepieces$village;
+@@ -1288,17 +1291,24 @@
+       public List<VillagePieces.PieceWeight> field_74931_h;
+       public List<StructurePiece> field_74932_i = Lists.<StructurePiece>newArrayList();
+       public List<StructurePiece> field_74930_j = Lists.<StructurePiece>newArrayList();
++      public @javax.annotation.Nullable net.minecraft.world.biome.Biome biome;
+ 
+       public Start() {
+       }
+ 
+       public Start(int p_i48769_1_, Random p_i48769_2_, int p_i48769_3_, int p_i48769_4_, List<VillagePieces.PieceWeight> p_i48769_5_, VillageConfig p_i48769_6_) {
++         this(p_i48769_1_, p_i48769_2_, p_i48769_3_, p_i48769_4_, p_i48769_5_, p_i48769_6_, null);
++      }
++      
++      public Start(int p_i48769_1_, Random p_i48769_2_, int p_i48769_3_, int p_i48769_4_, List<VillagePieces.PieceWeight> p_i48769_5_, VillageConfig p_i48769_6_, net.minecraft.world.biome.Biome biome) {
+          super((VillagePieces.Start)null, 0, p_i48769_2_, p_i48769_3_, p_i48769_4_);
+          this.field_74931_h = p_i48769_5_;
+          this.field_74928_c = p_i48769_6_.field_202461_a;
+          this.field_189928_h = p_i48769_6_.field_202462_b;
+          this.func_202579_a(this.field_189928_h);
+          this.field_189929_i = p_i48769_2_.nextInt(50) == 0;
++         this.biome = biome;
++         this.startPiece = this;
+       }
+    }
+ 
+@@ -1363,11 +1373,12 @@
+       }
+    }
+ 
+-   abstract static class Village extends StructurePiece {
++   public abstract static class Village extends StructurePiece {
+       protected int field_143015_k = -1;
+       private int field_74896_a;
+       protected VillagePieces.Type field_189928_h;
+       protected boolean field_189929_i;
++      protected VillagePieces.Start startPiece;
+ 
+       public Village() {
+       }
+@@ -1377,6 +1388,7 @@
+          if (p_i2107_1_ != null) {
+             this.field_189928_h = p_i2107_1_.field_189928_h;
+             this.field_189929_i = p_i2107_1_.field_189929_i;
++            this.startPiece = p_i2107_1_;
+          }
+ 
+       }
+@@ -1480,13 +1492,12 @@
```

Comment:
This patch needs updating to match #5200.

Reply:
I'm intentionally porting as-is so that it's easier to fast-forward the 1.13 branch when the time comes. Anything added to 1.12 after the branch was made is being left out for now.